### ISSUE TITLE
AutoMap in version 8.x

### ISSUE THREAD

**Issue by lukapor** [elastic/elasticsearch-net#6610]
My trying to rewrite my code with new elasticsearch client v8-beta.1.

I am using mapping attributes on index class and AutoMap function. I did not find this feature in new version.

I hope this will be possible.

Thank You in Advance.

**Comment by stevejgordon** [comment]
Hi, @lukapor. This is not in the current beta's, but we do plan to port the functionality before GA. It's always helpful to have examples of real usage that we can test against, so if you are able to share a code snippet of where you are using this, that would be great!

**Comment by lukapor** [comment]
Hi, I want to write a test but I saw that you allready have test for that.

7.17/tests/Tests/ClientConcepts/HighLevel/Mapping/AttributeMapping.doc.cs

**Comment by meriturva** [comment]
just to know...is it implemented now on 8.x version?

**Comment by stevejgordon** [comment]
@meriturva No, this didn't make the cut for the initial release. I don't have an ETA for us re-introducing it yet.

**Comment by ErroneousFatality** [comment]
@stevejgordon What would your advice be for my usecase: I want to use the inferred mapping for all properties except for one I want to set explicitly. I can't seem to find a way to do that through the .DefaultMappingFor method. Must I call `.CreateIndexAsync` and explicitly map out the whole object by hand?

**Comment by mwrothbe-intc** [comment]
I am also in need of mapping attributes in v8. My problem is in NEST v7, the densevector attribute is missing the index and similarity properties. It’s not clear to me how to me how to enable index=true and similarity=“dot_product” on new index mapping without those attribute properties. Assuming NEST v7 won’t be updated, please enable attribute mapping in v8 and include the index and similarity options for densevector.

**Comment by flobernd** [comment]
@mwrothbe-intc Manual mapping is already possible using the v8 client. This issue is only about the missing `AutoMap` functionality 🙂 

For your case, something like this should work:

```csharp
var mapResponse = client.Indices.PutMapping("index", x => x
    .Properties<Person>(p => p
        .DenseVector(x => x.Data, d => d
            .Index(true)
            .Similarity("dot_product"))));
```

**Comment by huey2531** [comment]
any idea when AutoMap will be added? still on v7, been waiting for this feature

**Comment by sbodese** [comment]
hi, 
i agree also, the AutoMap Feature is near a "must have", if the index creation is fully OO driven, otherwise there is only a manual way with a static json mapping file usage, or write a own AutoMapper Function, But this seems not so trivial, if i take a look on the old AutoMapper code by the client 7.x.
@stevejgordon what is your recommendation pls for a workaround about this missing AutoMap function?

**Comment by flobernd** [comment]
Steve is no longer maintaining the library (this is now my job). Until I reintroduce auto-mapping, your options are:

- Use the v7 NEST client in compatibility mode just for index creation (it can be used side-by-side with the new v8 client)
- Use manual mapping (like described in my previous comment)
- Write your own auto-mapper implementation based on the NEST code

That being said, I'm fully aware this is one of the more frequently requested features, but I can't give you any definitive ETA.

**Comment by sbodese** [comment]
Hi @flobernd 
i got some APi issues with the index creation by the nest 7.15.5 client, if i use the newer .map(...) api calls 

[https://github.com/elastic/elasticsearch-net/issues/7929](url)

i have solved this issues with the older .mappings(... ) APi usage.

**Comment by DR9885** [comment]
I tried to use old client to Automap, and it fails on enums.

**Comment by pschwarzpp** [comment]
> Hi, @lukapor. This is not in the current beta's, but we do plan to port the functionality before GA. It's always helpful to have examples of real usage that we can test against, so if you are able to share a code snippet of where you are using this, that would be great!

So it's now 2024 and the NEST client is deprecated in nuget. Is there an ETA on this functionality?

**Comment by Crazy1van** [comment]
I would like to know as well.  We just tried migrating away from NEST, which went smoothly until we encountered the lack of automap.

**Comment by niemyjski** [comment]
We really really need this. We have over 200 indices and this is in our core library.

**Comment by flobernd** [comment]
Hi 👋 Just to clarify a few things:

1. AutoMapping functionality **will** come back.
2. The current plan is to release this as an independant package to allow for re-use in some of our other projects. This might introduce some technical difficulties which we don't have discussed in detail yet.
3. There still is no ETA. I'm the only person working on the .NET client right now and besides that I'm as well responsible for other high priority projects.
4. I'm happy to accept PRs to kick-start this project if anybody is keen starting to port over the auto mapping code from NEST to the v8 client.

**Comment by niemyjski** [comment]
Is there any eta when this might be completed? We really really need this

**Comment by flobernd** [comment]
I know that this is probably the most awaited high level feature that I promised to port over from NEST, but I still don't have capacity to work on this right now.

Currently on the roadmap (ordered by priority from top to bot):

1. **Improve documentation**
    I'm currently working on a tool that automatically generates C# code snippets from (Kibana) REST/JSON request inputs. This allows us to display corresponding C# code for each example code on our REST API documentation.
2. **Improve usability**
    Planning and implementing a new descriptor/object design which solves multiple pain-points like e.g. the missing functionality to inspect the state of a descriptor or to convert between the descriptor/object representation.
    Evaluating various other changes to improve the quality and usability of the generated code (like exposing most fields publically, reverting from `init` to `set`, etc.).
3. **High level helpers**
    Like e.g. AutoMap.

My previous statements are still valid; especially:

> I'm happy to accept PRs to kick-start this project if anybody is keen starting to port over the auto mapping code from NEST to the v8 client.

**Comment by niemyjski** [comment]
I appreciate the roadmap. I kinda feel like documentation is the least of the concerns for those trying to upgrade from nest as there is a long list of issues (point 2 and 3).

Also is the NEST client getting an extension on being already deprecated as there are limited resources to fill feature gap used by many in the community? I truly wish I had more time and resources myself to do bigger pr's like AutoMap.

**Comment by niemyjski** [comment]
In the grand scheme of things, I feel like this should have the lowest priority. Production apps should be setting explicit mappings.

**Comment by huey2531** [comment]
any updates on this now that 9.0 is out?

**Comment by flobernd** [comment]
@huey2531 I don't expect having capacity to work on this in the next months. There are still quite a few high priority tasks for elasticsearch-net and as well other projects that I have to work on in parallel.

**Comment by muralidharanvenkataraman** [comment]
I am working with a client and they are looking for the support to auto feature.

**Comment by jmenagh** [comment]
> they are looking for the support to auto f

AutoMap is a critical feature that eliminates the need to manually define index mappings. We would greatly appreciate having this functionality restored as soon as possible, especially as we are actively migrating our Elasticsearch servers to the cloud.

In the absence of this feature, we are still relying on deprecated NuGet packages. Implementing a manual workaround to define index mappings would require additional time and resources—something we’re hoping to avoid during this transition.

**Comment by huey2531** [comment]
been waiting for 2 years now, since 2023!

**Comment by niemyjski** [comment]
I was in the boat that this would be really good to have, the more time goes on I wonder if it's even a good idea. Explicit mappings are a good thing and so is turning off dynamic and using dynamic templates if you really need it.

**Comment by huey2531** [comment]
@flobernd any updates on this would be great. Thanks!

**Comment by huey2531** [comment]
@flobernd any chance this will be implemented by the end of year 2025?

**Comment by flobernd** [comment]
We have not forgotten about this feature, but as usual, we can not give you a definitive ETA.

With AOT getting more and more popular, we can not rely on reflection anymore for this task (at least not in all cases). A custom code generator will be required.

**Comment by kmcclellan** [comment]
>Use the v7 NEST client in compatibility mode just for index creation (it can be used side-by-side with the new v8 client)

Is it fair to say this is no longer viable option when upgrading to an Elasticsearch v9 server? Only current option is to migrate to the Fluent API?

Like many, I have a LOT of fields mapped this way, so this is not something I want to do if the future recommended way will be to go back to attribute-driven code generation.

I suppose another option would be to copy the current mappings from the server and upload them directly (as JSON) using the transport client. I may try my hand at this and post if it works out.

**Comment by handloong** [comment]
I hope this will be possible in V10 😂

**Comment by flobernd** [comment]
We recently started working on this feature:

https://github.com/elastic/elastic-ingest-dotnet/tree/main/src/Elastic.Mapping

There is still quite a bit of work to do before this is production ready and we can integrate this in the client package.
